spacer/tests/test_extract_features.py

```diff
@@ -177,5 +178,133 @@ def test_regression(self):
+
+        # Check some feature metadata
+        self.assertEqual(features.point_features[0].row, 148)
+        self.assertEqual(features.point_features[0].col, 50)
+        self.assertEqual(len(features.point_features[0].data), 1280)
+
+    def test_corner_case2(self):
+        msg = ExtractFeaturesMsg(
+            job_token='cornercase_2',
+            feature_extractor_name='efficientnet_b0_ver1',
+            rowcols=[(190, 226), (25, 359)],
+            image_loc=DataLocation(storage_type='s3',
+                                   key='sfq2mr5qbs.jpeg',
+                                   bucket_name='spacer-test'),
+            feature_loc=DataLocation(storage_type='memory',
+                                     key='dummy')
+        )
+
+        ext = feature_extractor_factory(msg.feature_extractor_name)
+        img = load_image(msg.image_loc)
+        features, return_msg = ext(img, msg.rowcols)
+
+        self.assertTrue(isinstance(return_msg, ExtractFeaturesReturnMsg))
+        self.assertTrue(isinstance(features, ImageFeatures))
+
+        # Check some feature metadata
+        self.assertEqual(features.point_features[0].row, 190)
+        self.assertEqual(features.point_features[0].col, 226)
+        self.assertEqual(len(features.point_features[0].data), 1280)
+
+    def test_regression(self):
+        rowcols = [(20, 265),
+                   (76, 295),
+                   (59, 274),
+                   (151, 62),
+                   (265, 234)]
+
+        msg = ExtractFeaturesMsg(
+            job_token='regression_job',
+            feature_extractor_name='efficientnet_b0_ver1',
+            rowcols=rowcols,
+            image_loc=DataLocation(storage_type='s3',
+                                   key='08bfc10v7t.png',
+                                   bucket_name='spacer-test'),
+            feature_loc=DataLocation(storage_type='memory',
+                                     key='dummy')
+        )
+
+        legacy_feat_loc = DataLocation(storage_type='s3',
+                                       key='08bfc10v7t.png.effnet.featurevector',
+                                       bucket_name='spacer-test')
+
+        ext = feature_extractor_factory(msg.feature_extractor_name)
+
+        img = load_image(msg.image_loc)
+        features_new, _ = ext(img, msg.rowcols)
+        features_legacy = ImageFeatures.load(legacy_feat_loc)
+
+        self.assertFalse(features_legacy.valid_rowcol)
+        self.assertEqual(features_legacy.npoints, len(rowcols))
+        self.assertEqual(features_legacy.feature_dim, 1280)
+
+        self.assertTrue(features_new.valid_rowcol)
+        self.assertEqual(features_new.npoints, len(rowcols))
+        self.assertEqual(features_new.feature_dim, 1280)
+
+        for pf_new, pf_legacy in zip(features_new.point_features,
+                                     features_legacy.point_features):
+            self.assertTrue(np.allclose(pf_legacy.data, pf_new.data, rtol=1e-4))
```

Comment:
Also, you are right. This passes fine on my dev machine, but not on Travis CI. Maybe this was due to that `libjpeg` after all? Hmm. I suppose you could change to using `png` format instead.

Reply:
... actually, it passes fine now. Nevermind -- let's leave it as is for now.

Reply:
... actually, it does not pass. Can you try using `png`?

Reply:
actually the images I used for generating the legacy features are already `png`, and the original image for production testing is also `png` so technically it should work. Weird...

Reply:
This is so weird as I checked the array values of every step using my codebase and spacer and they are all the same: original images value, cropped patches value, before and after pre-processing value, all with `png` format. And I did turn off the BN and dropout otherwise the results will be changed every time the extractor is called. The thing is it works perfectly in local...

Reply:
I think I fixed it. Turns out you were setting the wrong parameter of `np.allclose`. This is the formula: `absolute(arr1 - arr2) <= (atol + rtol * absolute(arr2))`. You were setting `rtol` < 1 which made the criterion even more strict. I set `atol=1e-4` and that did the trick.

Reply:
(I'm still seeing problems related to lag connecting to S3 from Travis, but that is a different story, usually works after a restart).

Reply:
Great!! BTW EfficientNet-b0 training on dataset V3 will be finished in a few hours, I will use this legendary for evaluation.